napari/layers/points/_slice.py

```diff
@@ -0,0 +1,98 @@
+from dataclasses import dataclass, field
+from typing import Any
+
+import numpy as np
+
+from napari.layers.utils._slice_input import _SliceInput
+
+
+@dataclass(frozen=True)
+class _PointSliceResponse:
+    """Contains all the output data of slicing an image layer.
+
+    Attributes
+    ----------
+    indices : array like
+        Indices of the sliced Points data.
+    scale: array like or none
+    """
+
+    indices: np.ndarray = field(repr=False)
+    scale: Any = field(repr=False)
+
+
+@dataclass(frozen=True)
+class _PointSliceRequest:
+    """A callable that stores all the input data needed to slice a Points layer.
+
+    This should be treated a deeply immutable structure, even though some
+    fields can be modified in place. It is like a function that has captured
+    all its inputs already.
+
+    In general, the calling an instance of this may take a long time, so you may
+    want to run it off the main thread.
+
+    Attributes
+    ----------
+    dims : _SliceInput
+        Describes the slicing plane or bounding box in the layer's dimensions.
+    data : Any
+        The layer's data field, which is the main input to slicing.
+    dims_indices : Tuple[Union[int, slice], ...]
+        The slice indices in the layer's data space.
+    size : array like
+        Size of each point. This is used in calculating visibility.
+    others
+        See the corresponding attributes in `Layer` and `Image`.
+    """
+
+    dims: _SliceInput
+    data: Any = field(repr=False)
+    dims_indices: Any = field(repr=False)
+    size: Any = field(repr=False)
+    out_of_slice_display: bool = field(repr=False)
+
+    def __call__(self) -> _PointSliceResponse:
+        # Return early if no data
+        if len(self.data) == 0:
+            return _PointSliceResponse(indices=[], scale=np.empty(0))
+
+        not_disp = list(self.dims.not_displayed)
```

Comment:
If `not_disp` is empty, doesn't this imply that indices is just `range(len(self.data))`? And I think `scale` would be 1. Maybe another simple optimization (over `main`) that would help a little in some cases for 2D display of 2D points, and 3D display of 3D points.

Doesn't need to happen in this PR, but seems worth it to me.

Reply:
I tried and things broke, but I didn't try to figure out why it broke.

Reply:
Based on the profile dump, I think this is actually the cause of the performance regression because we hit this case there and you can see most of the time comes from `np.array`.

Using `np.arange(len(self.data), dtype=int)` seems to fix it, which is also consistent with %timeit:

```python
In [7]: %timeit a = np.array(range(65536), dtype=int)
3.25 ms ± 38.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [8]: %timeit a = np.arange(65536, dtype=int)
18.2 µs ± 247 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```